File: app/routers/postgresql.py
from fastapi import APIRouter, HTTPException, Depends
from pydantic import BaseModel, Field, validator
from typing import Dict, Any, List, Optional, Union
import logging
import psycopg2
from psycopg2.extras import RealDictCursor
import psycopg2.errors
import re
from app.utils.response import success_response, error_response
from app.utils.pool import postgresql_pool

router = APIRouter(tags=["PostgreSQL"])
logger = logging.getLogger("database-api")

class PostgreSQLConnectionInfo(BaseModel):
    host: str
    port: int = 5432
    database: str
    user: str
    password: str
    sslmode: Optional[str] = "prefer"
    connect_timeout: Optional[int] = 30

class PostgreSQLExecuteRequest(BaseModel):
    connection: PostgreSQLConnectionInfo
    sql: str
    parameters: Optional[Union[List[Any], Dict[str, Any]]] = []

    @validator('sql')
    def validate_sql(cls, v):
        """基本的SQL语法验证"""
        if not v or not v.strip():
            raise ValueError("SQL语句不能为空")
        
        # 检查常见SQL语法错误
        sql_lower = v.lower().strip()
        
        # 检查SELECT语句是否有FROM子句
        if sql_lower.startswith('select'):
            if ' from ' not in sql_lower:
                raise ValueError("SELECT语句缺少FROM子句")
            
            # 检查是否指定了列
            select_from_parts = sql_lower.split(' from ')[0]
            if select_from_parts == 'select' or select_from_parts.strip() == 'select':
                raise ValueError("SELECT语句缺少列名，请指定要查询的列或使用 SELECT * 查询所有列")
            
            # 检查select和from之间是否有非空白字符
            select_part_match = re.match(r'select\s+from', sql_lower)
            if select_part_match:
                raise ValueError("SELECT语句缺少列名，请指定要查询的列或使用 SELECT * 查询所有列")
            
        # 检查基本的括号匹配
        if v.count('(') != v.count(')'):
            raise ValueError("SQL语句括号不匹配")
            
        # 检查WHERE后是否有条件
        where_pattern = re.compile(r'where\s+;', re.IGNORECASE)
        if where_pattern.search(v):
            raise ValueError("WHERE子句后缺少条件")
            
        # 检查常见的拼写错误
        common_typos = {
            r'\bslect\b': 'select', 
            r'\bform\b': 'from',
            r'\bwhere\s+and\b': 'where',
            r'\bgroup\s+order\b': 'group by ... order'
        }
        
        for typo, correction in common_typos.items():
            if re.search(typo, sql_lower):
                raise ValueError(f"SQL语句可能存在拼写错误: 检查 '{correction}'")
                
        return v

@router.post("/postgresql")
async def execute_postgresql(request: PostgreSQLExecuteRequest):
    """
    执行PostgreSQL SQL语句
    
    - 支持查询、插入、更新、删除操作
    - 返回查询结果或影响的行数
    """
    conn = None
    try:
        # 额外的SQL语法验证
        sql_lower = request.sql.lower().strip()
        
        # 特别检查SELECT语句是否包含列指定
        if sql_lower.startswith('select'):
            select_parts = sql_lower.split(' from ')[0].strip()
            if select_parts == 'select' or re.match(r'select\s*$', select_parts):
                return error_response(1005, "SQL验证错误: SELECT语句缺少列名，请指定要查询的列或使用 SELECT * 查询所有列")
        
        # 从连接池获取连接
        conn, error = postgresql_pool.get_connection(
            host=request.connection.host,
            port=request.connection.port,
            database=request.connection.database,
            user=request.connection.user,
            password=request.connection.password,
            sslmode=request.connection.sslmode,
            connect_timeout=request.connection.connect_timeout
        )
        
        if error:
            logger.error(f"从连接池获取连接失败: {error}")
            return error_response(1001, f"数据库连接错误: {error}")
        
        logger.info(f"成功从连接池获取连接: {request.connection.host}:{request.connection.port}/{request.connection.database}")
        
        # 创建游标，使用RealDictCursor返回字典格式结果
        with conn.cursor(cursor_factory=RealDictCursor) as cursor:
            # 执行SQL
            logger.info(f"执行SQL: {request.sql}")
            
            try:
                cursor.execute(request.sql, request.parameters)
            except psycopg2.ProgrammingError as e:
                # 捕获SQL执行错误，提供更详细的错误信息
                error_msg = str(e)
                
                # 尝试提取更有用的错误信息
                if "column" in error_msg and "does not exist" in error_msg:
                    match = re.search(r'column "([^"]+)" does not exist', error_msg)
                    if match:
                        column = match.group(1)
                        error_msg = f"列 '{column}' 不存在，请检查列名是否正确或表是否存在此列"
                elif "missing FROM-clause" in error_msg:
                    error_msg = "SQL语句缺少FROM子句，SELECT语句需要指定查询的表"
                elif "syntax error at or near" in error_msg:
                    match = re.search(r'syntax error at or near "([^"]+)"', error_msg)
                    if match:
                        problematic_part = match.group(1)
                        error_msg = f"SQL语法错误，问题出现在: '{problematic_part}'"
                elif "SELECT列表中的表达式" in error_msg or "column reference" in error_msg or "no columns specified" in error_msg:
                    error_msg = "SELECT语句缺少列名，请指定要查询的列或使用 SELECT * 查询所有列"
                
                logger.error(f"SQL执行错误: {error_msg}")
                return error_response(1003, f"SQL执行错误: {error_msg}")
            
            # 检查结果是否为空
            sql_type = request.sql.strip().upper().split(None, 1)[0]
            
            if sql_type in ("SELECT", "WITH", "SHOW", "EXPLAIN"):
                # 查询操作，返回结果集
                result = cursor.fetchall()
                
                # 将结果转换为普通字典列表
                result_list = []
                
                for row in result:
                    row_dict = {}
                    # RealDictRow实际上是类似字典的对象，可以直接迭代其key-value对
                    for key in row.keys():
                        value = row[key]
                        # 处理特殊数据类型
                        if hasattr(value, 'isoformat'):  # 日期/时间类型
                            row_dict[key] = value.isoformat()
                        elif isinstance(value, (int, float, str, bool, type(None))):
                            row_dict[key] = value
                        else:
                            row_dict[key] = str(value)  # 其他类型转为字符串
                    
                    result_list.append(row_dict)
                
                # 检查结果是否为空
                if not result_list:
                    logger.info("SQL查询未返回任何结果")
                
                return success_response(result_list)
            else:
                # 非查询操作，提交事务并返回影响行数
                conn.commit()
                affected_rows = cursor.rowcount
                
                # 检查影响的行数
                if affected_rows == 0:
                    logger.info("SQL操作未影响任何行")
                
                return success_response(affected_rows)
                
    except psycopg2.OperationalError as e:
        logger.error(f"PostgreSQL连接错误: {str(e)}")
        return error_response(1001, f"数据库连接错误: {str(e)}")
    except psycopg2.errors.UndefinedTable as e:
        # 提取表名
        error_msg = str(e)
        match = re.search(r'relation "([^"]+)" does not exist', error_msg)
        if match:
            table_name = match.group(1)
            error_msg = f"表 '{table_name}' 不存在"
        
        logger.error(f"表不存在: {error_msg}")
        return error_response(1002, f"表不存在: {error_msg}")
    except psycopg2.errors.SyntaxError as e:
        error_msg = str(e)
        
        # 特殊处理SELECT语句缺少列
        if "SELECT" in error_msg and ("column reference" in error_msg or "target lists" in error_msg):
            error_msg = "SELECT语句缺少列名，请指定要查询的列或使用 SELECT * 查询所有列"
        
        logger.error(f"SQL语法错误: {error_msg}")
        return error_response(1003, f"SQL语法错误: {error_msg}")
    except psycopg2.errors.InFailedSqlTransaction as e:
        logger.error(f"SQL事务失败: {str(e)}")
        if conn:
            conn.rollback()
        return error_response(1004, f"SQL事务失败: {str(e)}")
    except psycopg2.Error as e:
        logger.error(f"PostgreSQL错误: {str(e)}")
        if conn:
            conn.rollback()
        return error_response(1000, f"数据库错误: {str(e)}")
    except ValueError as e:
        # 处理SQL验证错误
        logger.error(f"SQL验证错误: {str(e)}")
        return error_response(1005, f"SQL验证错误: {str(e)}")
    except Exception as e:
        logger.error(f"未预期的错误: {str(e)}")
        if conn:
            conn.rollback()
        return error_response(9999, f"服务器错误: {str(e)}")
    finally:
        # 释放连接回连接池，而不是关闭连接
        if conn:
            try:
                postgresql_pool.release_connection(
                    host=request.connection.host, 
                    port=request.connection.port, 
                    database=request.connection.database, 
                    user=request.connection.user, 
                    conn=conn
                )
                logger.info("数据库连接已归还到连接池")
            except Exception as e:
                logger.error(f"归还连接到连接池失败: {str(e)}")
                try:
                    conn.close()
                    logger.info("无法归还到连接池，已关闭连接")
                except:
                    pass 

File: controllers/postgresql.go
package controllers

import (
	"database/sql"
	"fmt"
	"regexp"
	"strings"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/lib/pq"
	"github.com/renoelis/database-api-go/models"
	"github.com/renoelis/database-api-go/services"
	"github.com/renoelis/database-api-go/utils"
	"github.com/sirupsen/logrus"
)

// ExecutePostgreSQL 处理PostgreSQL SQL执行请求
func ExecutePostgreSQL(c *gin.Context) {
	logger := logrus.New()

	// 绑定请求数据
	var request models.PostgreSQLExecuteRequest
	if err := c.ShouldBindJSON(&request); err != nil {
		logger.Errorf("解析PostgreSQL请求失败: %v", err)
		utils.ResponseError(c, 400, 1000, fmt.Sprintf("请求参数错误: %v", err))
		return
	}

	// 验证SQL语句语法
	if err := request.ValidateSQL(); err != nil {
		logger.Errorf("SQL验证错误: %v", err)
		utils.ResponseError(c, 400, 1005, fmt.Sprintf("SQL验证错误: %v", err))
		return
	}

	// 额外的SQL语法验证
	sqlLower := strings.ToLower(strings.TrimSpace(request.SQL))
	if strings.HasPrefix(sqlLower, "select") {
		selectParts := strings.Split(sqlLower, " from ")[0]
		if selectParts == "select" || regexp.MustCompile(`select\s*$`).MatchString(selectParts) {
			utils.ResponseError(c, 400, 1005, "SQL验证错误: SELECT语句缺少列名，请指定要查询的列或使用 SELECT * 查询所有列")
			return
		}
	}

	// 获取连接池实例
	pgPool := services.GetPostgreSQLInstance()

	// 从连接池获取连接
	conn, err := pgPool.GetConnection(
		request.Connection.Host,
		request.Connection.Port,
		request.Connection.Database,
		request.Connection.User,
		request.Connection.Password,
		request.Connection.SSLMode,
		request.Connection.ConnectTimeout,
	)

	if err != nil {
		logger.Errorf("从连接池获取连接失败: %v", err)
		utils.ResponseError(c, 500, 1001, fmt.Sprintf("数据库连接错误: %v", err))
		return
	}

	logger.Infof("成功从连接池获取连接: %s:%d/%s", 
		request.Connection.Host, 
		request.Connection.Port, 
		request.Connection.Database)

	// 执行SQL
	logger.Infof("执行SQL: %s", request.SQL)
	
	// 处理不同类型的SQL语句
	sqlType := strings.ToUpper(strings.Fields(strings.TrimSpace(request.SQL))[0])

	// 如果是查询操作
	if sqlType == "SELECT" || sqlType == "WITH" || sqlType == "SHOW" || sqlType == "EXPLAIN" {
		// 查询数据
		rows, err := executeQueryWithParams(conn, request.SQL, request.Parameters)
		if err != nil {
			handleSQLError(c, err)
			return
		}
		defer rows.Close()

		// 转换结果为JSON
		result, err := convertRowsToJSON(rows)
		if err != nil {
			logger.Errorf("转换结果集失败: %v", err)
			utils.ResponseError(c, 500, 1006, fmt.Sprintf("处理查询结果失败: %v", err))
			return
		}

		// 检查结果是否为空
		if len(result) == 0 {
			logger.Info("SQL查询未返回任何结果")
		}

		// 返回结果
		utils.ResponseSuccess(c, result)
	} else {
		// 执行非查询操作
		result, err := executeNonQueryWithParams(conn, request.SQL, request.Parameters)
		if err != nil {
			handleSQLError(c, err)
			return
		}

		// 检查影响的行数
		if result == 0 {
			logger.Info("SQL操作未影响任何行")
		}

		// 返回影响行数
		utils.ResponseSuccess(c, result)
	}

	// 释放连接
	pgPool.ReleaseConnection(
		request.Connection.Host,
		request.Connection.Port,
		request.Connection.Database,
		request.Connection.User,
	)
	logger.Info("数据库连接已归还到连接池")
}

// executeQueryWithParams 执行带参数的查询
func executeQueryWithParams(db *sql.DB, query string, params interface{}) (*sql.Rows, error) {
	// 根据参数类型执行不同的查询
	switch p := params.(type) {
	case []interface{}:
		// 数组参数
		return db.Query(query, p...)
	case map[string]interface{}:
		// 命名参数，需要转换
		return executeNamedQuery(db, query, p)
	default:
		// 无参数
		return db.Query(query)
	}
}

// executeNonQueryWithParams 执行带参数的非查询
func executeNonQueryWithParams(db *sql.DB, query string, params interface{}) (int64, error) {
	var result sql.Result
	var err error

	switch p := params.(type) {
	case []interface{}:
		// 数组参数
		result, err = db.Exec(query, p...)
	case map[string]interface{}:
		// 命名参数，需要转换
		result, err = executeNamedExec(db, query, p)
	default:
		// 无参数
		result, err = db.Exec(query)
	}

	if err != nil {
		return 0, err
	}

	return result.RowsAffected()
}

// executeNamedQuery 执行命名参数查询
func executeNamedQuery(db *sql.DB, query string, params map[string]interface{}) (*sql.Rows, error) {
	// 这里需要将命名参数转换为位置参数
	// 简单实现，生产环境应使用更健壮的实现
	var args []interface{}
	for _, v := range params {
		args = append(args, v)
	}
	return db.Query(query, args...)
}

// executeNamedExec 执行命名参数非查询
func executeNamedExec(db *sql.DB, query string, params map[string]interface{}) (sql.Result, error) {
	// 同上，简单实现
	var args []interface{}
	for _, v := range params {
		args = append(args, v)
	}
	return db.Exec(query, args...)
}

// convertRowsToJSON 将SQL结果转换为JSON格式
func convertRowsToJSON(rows *sql.Rows) ([]map[string]interface{}, error) {
	// 获取列信息
	columns, err := rows.Columns()
	if err != nil {
		return nil, err
	}

	// 创建结果集
	var result []map[string]interface{}

	// 预分配值接收对象
	values := make([]interface{}, len(columns))
	valuePtrs := make([]interface{}, len(columns))
	for i := range columns {
		valuePtrs[i] = &values[i]
	}

	// 逐行处理结果
	for rows.Next() {
		// 扫描当前行到对象
		if err := rows.Scan(valuePtrs...); err != nil {
			return nil, err
		}

		// 创建一个新的行对象
		row := make(map[string]interface{})

		// 填充行数据
		for i, col := range columns {
			var v interface{}
			val := values[i]

			// 检查值类型，并适当转换
			switch value := val.(type) {
			case nil:
				v = nil
			case []byte:
				// 尝试转换为字符串
				v = string(value)
			case time.Time:
				// 格式化时间为ISO字符串
				v = value.Format(time.RFC3339)
			default:
				v = value
			}

			row[col] = v
		}

		// 添加到结果集
		result = append(result, row)
	}

	// 检查行遍历是否有错误
	if err := rows.Err(); err != nil {
		return nil, err
	}

	return result, nil
}

// handleSQLError 处理SQL执行错误
func handleSQLError(c *gin.Context, err error) {
	logger := logrus.New()
	logger.Errorf("SQL执行错误: %v", err)

	// 处理不同类型的PostgreSQL错误
	if pqErr, ok := err.(*pq.Error); ok {
		// 根据PostgreSQL错误代码返回不同的错误消息
		switch pqErr.Code {
		case "42P01": // 表不存在
			// 提取表名
			match := regexp.MustCompile(`relation "([^"]+)" does not exist`).FindStringSubmatch(pqErr.Message)
			if len(match) > 1 {
				tableName := match[1]
				utils.ResponseError(c, 400, 1002, fmt.Sprintf("表不存在: %s", tableName))
				return
			}
			utils.ResponseError(c, 400, 1002, fmt.Sprintf("表不存在: %v", pqErr.Message))
		case "42601": // 语法错误
			utils.ResponseError(c, 400, 1003, fmt.Sprintf("SQL语法错误: %v", pqErr.Message))
		case "42703": // 列不存在
			match := regexp.MustCompile(`column "([^"]+)" does not exist`).FindStringSubmatch(pqErr.Message)
			if len(match) > 1 {
				column := match[1]
				utils.ResponseError(c, 400, 1003, fmt.Sprintf("列 '%s' 不存在，请检查列名是否正确或表是否存在此列", column))
				return
			}
			utils.ResponseError(c, 400, 1003, fmt.Sprintf("SQL执行错误: %v", pqErr.Message))
		case "25P02": // 事务失败
			utils.ResponseError(c, 400, 1004, fmt.Sprintf("SQL事务失败: %v", pqErr.Message))
		default:
			utils.ResponseError(c, 400, 1000, fmt.Sprintf("数据库错误: %v", pqErr.Message))
		}
	} else {
		utils.ResponseError(c, 400, 1000, fmt.Sprintf("数据库错误: %v", err.Error()))
	}
} 